### ISSUE TITLE
Spurious failures when iterating over an empty vector

### ISSUE THREAD

**Issue by zhassan-aws** [model-checking/kani#763]
I tried this code:

```rust
struct Foo (u64);

fn main() {
    let x = Vec::<Foo>::new();
    for i in x {}
}
```

using the following command line invocation:

```
kani test.rs
```

with Kani version: `8f3d0911697`

I expected to see this happen: Verification successful

Instead, this happened: 15 checks failed:
```
[pointer_arithmetic.7] pointer relation: pointer NULL in var_11: FAILURE
[pointer_arithmetic.9] pointer relation: deallocated dynamic object in var_11: FAILURE
[pointer_arithmetic.10] pointer relation: dead object in var_11: FAILURE
[pointer_arithmetic.11] pointer relation: pointer outside object bounds in var_11: FAILURE
[pointer_arithmetic.12] pointer relation: invalid integer address in var_11: FAILURE
[pointer_arithmetic.13] pointer relation: pointer NULL in var_12: FAILURE
[pointer_arithmetic.15] pointer relation: deallocated dynamic object in var_12: FAILURE
[pointer_arithmetic.16] pointer relation: dead object in var_12: FAILURE
[pointer_arithmetic.17] pointer relation: pointer outside object bounds in var_12: FAILURE
[pointer_arithmetic.18] pointer relation: invalid integer address in var_12: FAILURE
[pointer_arithmetic.19] pointer arithmetic: pointer NULL in var_4 + var_6: FAILURE
[pointer_arithmetic.21] pointer arithmetic: deallocated dynamic object in var_4 + var_6: FAILURE
[pointer_arithmetic.22] pointer arithmetic: dead object in var_4 + var_6: FAILURE
[pointer_arithmetic.23] pointer arithmetic: pointer outside object bounds in var_4 + var_6: FAILURE
[pointer_arithmetic.24] pointer arithmetic: invalid integer address in var_4 + var_6: FAILURE
...
** 15 of 291 failed (3 iterations)
VERIFICATION FAILED
```
The same checks fail if I change to a normal struct:
```rust
struct Foo {
    x: u64,
}
```

**Comment by zhassan-aws** [comment]
This actually occurs without structs as well:
```
$ cat test.rs 
fn main() {
    let x = Vec::<i32>::new();
    for i in x {}
}
$ kani test.rs | grep FAIL
[pointer_arithmetic.13] pointer arithmetic: pointer NULL in var_4 + var_6: FAILURE
[pointer_arithmetic.15] pointer arithmetic: deallocated dynamic object in var_4 + var_6: FAILURE
[pointer_arithmetic.16] pointer arithmetic: dead object in var_4 + var_6: FAILURE
[pointer_arithmetic.17] pointer arithmetic: pointer outside object bounds in var_4 + var_6: FAILURE
[pointer_arithmetic.18] pointer arithmetic: invalid integer address in var_4 + var_6: FAILURE
[pointer_arithmetic.25] pointer relation: pointer NULL in var_11: FAILURE
[pointer_arithmetic.27] pointer relation: deallocated dynamic object in var_11: FAILURE
[pointer_arithmetic.28] pointer relation: dead object in var_11: FAILURE
[pointer_arithmetic.29] pointer relation: pointer outside object bounds in var_11: FAILURE
[pointer_arithmetic.30] pointer relation: invalid integer address in var_11: FAILURE
[pointer_arithmetic.31] pointer relation: pointer NULL in var_12: FAILURE
[pointer_arithmetic.33] pointer relation: deallocated dynamic object in var_12: FAILURE
[pointer_arithmetic.34] pointer relation: dead object in var_12: FAILURE
[pointer_arithmetic.35] pointer relation: pointer outside object bounds in var_12: FAILURE
[pointer_arithmetic.36] pointer relation: invalid integer address in var_12: FAILURE
VERIFICATION FAILED
```

**Comment by celinval** [comment]
I did some digging into this issue. I think there is a mix here of a valid failure and a CBMC usability issue. So far I investigated these failures:

```
[pointer_arithmetic.13] pointer arithmetic: pointer NULL in var_4 + var_6: FAILURE
[pointer_arithmetic.15] pointer arithmetic: deallocated dynamic object in var_4 + var_6: FAILURE
[pointer_arithmetic.16] pointer arithmetic: dead object in var_4 + var_6: FAILURE
[pointer_arithmetic.17] pointer arithmetic: pointer outside object bounds in var_4 + var_6: FAILURE
[pointer_arithmetic.18] pointer arithmetic: invalid integer address in var_4 + var_6: FAILURE
```

There is a legit issue caused because `std::Vec` implementation uses a dangling pointer to represent 0 capacity vectors and using `ptr::add()`. See [this discussion](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Behavior.20of.20dangling.20pointer.20arithmetic.20operations) for more context. Note that we might want to consider tuning Kani to not emit errors when user uses `unwrap_*()`, since the operation is safe even with dangling pointers.

Some of these failures are spurious though, e.g.: `pointer NULL in <expr>`. I believe the issue here is that because the pointer is dangling, CBMC treats the result of the addition as UB and all the subsequent checks fail even though `var_6` is `0` and `var_4` is not `NULL`.
